### PR TITLE
guide: Comparing Experiments

### LINKED ISSUE #2737: guide: Comparing Experiments
It may be worthwhile to have a section in the Experiment Management about `dvc exp show`, `dvc exp diff` and their options. 

Related #2548 

- [x] Is https://github.com/iterative/dvc.org/issues/2844 still needed after this?

### LINKED ISSUE #2860: guide: `--json/csv` examples in Listing/Comparing Experiments
I'll add an example with `jq` to the UG after testing it. 

_Originally posted by @iesahin in https://github.com/iterative/dvc.org/issues/2740#issuecomment-929304537_

### PR DESCRIPTION
**Pull request by iesahin**
Adds a section to the UG about `dvc exp list`, `dvc exp show` and `dvc exp diff`. 

Closes #2737 
Fixes #2860 

- [ ] Let's move unresolved discussions to https://github.com/iterative/dvc.org/issues/2995

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
@iesahin FYI please notice https://github.com/iterative/dvc.org/pull/2877.

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+└───────────────────────┴────────┴──────────────────┘
+```
+
+## Get experiments table in JSON
+
+There may be times when a machine readable format for the experiments is
+required. In its default settings `dvc exp show` lists the experiments in a
+visually appealing way that may not be suitable to parse in scripts. To get a
+list of experiments with their params and metrics, you can use `--show-json` or
+`--show-csv` flags.
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:245]: It'll be `--json` and `--cvs` soon (as hinted in https://github.com/iterative/dvc.org/pull/2862#issuecomment-932408692) ☝️

**Comment by jorgeorpinel** [comment]
Please add https://github.com/iterative/dvc.org/pull/2908#pullrequestreview-775523131 here.

**Comment by jorgeorpinel** [comment]
> Could you point to the place where these regexes are?

@iesahin https://github.com/iterative/link-check/blob/master/src/scrapeLinks.ts

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+{
+  "old": 0.9150999784469604,
+  "new": 0.9153000116348267,
+  "diff": 0.00020003318786621094
+}
+```
+
+### Get the comparison in Markdown table format
+
+`dvc exp diff` has also an option to output a [Github-Flavored Markdown][gfm]
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:484]: To test out https://github.com/iterative/dvc.org/pull/2862#issuecomment-954083610 ⌛ 
```suggestion

```

Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md]: Yep, this is the line that makes the link checker fail. I removed it and it's passing now. Will re-add it though (we can merge as-is)

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+{
+  "old": 0.9150999784469604,
+  "new": 0.9153000116348267,
+  "diff": 0.00020003318786621094
+}
+```
+
+### Get the comparison in Markdown table format
+
+table to embed in the reports directly.
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:484]: Re-add line that makes the link checker fail
```suggestion
`dvc exp diff` has also an option to output a [Github-Flavored Markdown][gfm]
table to embed in the reports directly.
```

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+...
+```
+
+[csvkit]: https://csvkit.readthedocs.io/en/latest/
+
+## Compare specific experiments
+
+In addition to showing a summary table of experiments, DVC provides a command to
+compare pairwise experiments. `dvc exp diff` is used to compare the experiments
+by the change in their metrics and params.
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:363]: > If we reorder the sections

BTW I thought we were doing that but I guess not (for now) so just this:

```suggestion
In addition to showing a summary table of experiments, DVC provides the
`dvc exp diff` command to compare pairs of experiments by the difference
in their metrics and params.
```

[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:363]: > committing along with a few uncontested copy edits ⌛

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+Path         Param             Value    Change
+params.yaml  model.conv_units  256      240
+params.yaml  train.epochs      10       0
+```
+
+## Compare an experiment with the workspace
+
+When you want to compare two experiments, either the baseline experiment in a
+commit, branch, tag or an attached experiment with ID, you can supply their
+names to `dvc exp diff`.
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:396]: If this is about comparing vs. workspace, why are we supplying 2 experiments to `exp diff` below? The only way to reliably compare the workspace (to `HEAD` aka baseline) is to use plan `exp diff` without arguments, no?

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+
+```
+$ dvc exp diff exp-25a26 cnn-64
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:399]: Let's just use another existing one for consistency for now? It's hypothetical anyway
```suggestion
$ dvc exp diff cnn-128 cnn-64
```

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+loss      0.23     -0.0048
+
+Param             Value    Change
+model.conv_units  64       -192
+```
+
+### Get the comparison in JSON
+
+Parsing the `dvc exp diff` output may not be feasible due to the custom table
+layout. When you want to use the output in other commands, `dvc exp diff` can
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:428]: 💅 "custom" sounds like user customized something but we're referring to the output of `exp diff` in general.
```suggestion
Parsing the output of `dvc exp diff` may not be easy when
you want to use it in other commands. `dvc exp diff` can
```

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+listed as keys.
+
+As an example, we can get only a specific metric [jq]:
+
+```dvc
+$ dvc exp diff exp-25a26 cnn-64 --json | jq '.metrics."metrics.json".acc'
+```
+
+```json
+{
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:475]: CE
```suggestion
As an example, we can get only a specific metric with [jq]:

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json | jq '.metrics."metrics.json".acc'
{
```

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+      }
+    }
+  }
+}
+```
+
+The output is a JSON dictionary with two keys, `metrics` and `params`, which
+have dictionaries as values. `metrics` and `params` dictionaries has keys for
+each of the metrics or params file, and for each file metrics and parameters are
+listed as keys.
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:466]: Thinking about it again, I don't think we need to document these output schemas that formally. Comprehensive examples (we got those already) should be enough.

**Review by jorgeorpinel — APPROVED** [review]
LGTM 🚀

File: content/docs/user-guide/experiment-management/comparing-experiments.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:394]: Agreed with @jorgeorpinel. The headers of these two sections seem reversed from the examples provided. "Compare specific experiments" shows a comparison to the workspace, and "Compare an experiment with the workspace" shows comparing two experiment revisions (not the workspace).

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+in a Git remote.
+
+```dvc
+$ dvc exp list origin
+refs/tags/baseline-experiment:
+        cnn-32
+        cnn-64
+```
+
+This command lists the experiments originated from `HEAD`. You can add any other
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/comparing-experiments.md:45]: @iesahin another question here... do you know how `HEAD` is determined for a remote to list experiments from? Is it the local repo's HEAD? Is it the default branch HEAD always? We should probably clarify (let's add it to https://github.com/iterative/dvc.org/issues/2995).

Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:45]: p.s. actually we could add it in https://github.com/iterative/dvc.org/pull/2908 as it's more about sharing than about comparing.

Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:45]: Moved to https://github.com/iterative/dvc.org/pull/2908#pullrequestreview-798148999.

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+
+## List experiments in the workspace
+
+After running the experiments, you can get a list of them by `dvc exp list`.
+Without any options, `dvc exp list` lists the experiments after the most recent
+commit.
+
+```dvc
+$ dvc exp list
+refs/tags/baseline-experiment:
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:15]: WDYT here @dberenbaum ? Rel. the "baseline" concept/term in general. Thanks

File: content/docs/user-guide/experiment-management/comparing-experiments.md [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+        cnn-96
+b99a33e:
+        exp-93150
+```
+
+## List experiments in another Git remote
+
+As we discuss in [Sharing Experiments], you can use `dvc exp push` to upload
+experiments to Git remotes. `dvc exp list` can be used to list the experiments
+in a Git remote.
```
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:36]: p.s. this is going away in https://github.com/iterative/dvc.org/pull/2908 anyway 🤷. Resolving

File: content/docs/user-guide/experiment-management/comparing-experiments.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/docs/user-guide/experiment-management/comparing-experiments.md:15]: I don't have strong opinion. Baseline makes sense to me, and I can see the argument for and against calling it an experiment. Calling it a baseline experiment seems fine. Having `baseline-experiment` as the tag feels a bit long and artificial, but again it's not a strong opinion. Seems like we have enough other areas where we can improve our experiments documentation that it's not worth spending too much time debating this. Hopefully we will get more crisp in how to explain experiments over time, seeing what works and what causes confusion.

**Pull request merged**